core/kernel/msg_param.c

```diff
@@ -31,6 +31,126 @@
 #include <kernel/msg_param.h>
 #include <mm/mobj.h>
 
+/**
+ * msg_param_extract_pages() - extract list of pages from
+ * OPTEE_MSG_ATTR_NONCONTIG buffer.
+ *
+ * @buffer:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_pages:  number of pages in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ *
+ * @buffer points to the physical address of a structure that can be viewed as
+ *
+ * struct page_data {
+ *   uint64_t pages_array[OPTEE_MSG_NONCONTIG_PAGE_SIZE/sizeof(uint64_t) - 1];
+ *   uint64_t next_page_data;
+ * };
+ *
+ * So, it is a linked list of arrays, where each element of linked list fits
+ * exactly into one 4K page.
+ *
+ * This function extracts data from arrays into one array pointed by @pages
+ *
+ * @buffer points to data shared with normal world, so some precautions
+ * should be taken.
+ */
+static bool msg_param_extract_pages(paddr_t buffer, paddr_t *pages,
+				       size_t num_pages)
+{
+	size_t cnt;
+	struct mobj *mobj;
+	paddr_t page;
+	uint64_t *va;
+	bool ret = false;
+
+	if (buffer & SMALL_PAGE_MASK)
+		return false;
+
+	/*
+	 * There we map first page of array.
+	 * mobj_mapped_shm_alloc() will check if page resides in nonsec ddr
+	 */
+	mobj = mobj_mapped_shm_alloc(&buffer, 1, 0, 0);
+	if (!mobj)
+		return false;
+
+	va = mobj_get_va(mobj, 0);
+	assert(va);
+
+	for (cnt = 0; cnt < num_pages; cnt++, va++) {
+		/*
+		 * If we about to roll over page boundary, then last entry holds
+		 * address of next page of array. Unmap current page and map
+		 * next one
+		 */
+		if (!((vaddr_t)(va + 1) & SMALL_PAGE_MASK)) {
```

Comment:
nitpicking: the use of `goto out` or `return false` with added comment looks a bit weird, even if fully functional.
Proposal:
```c
	page = *va;
	if (page & SMALL_PAGE_MASK)
		goto out;

	mobj_free(mobj);
	mobj = mobj_mapped_shm_alloc(&page, 1, 0, 0);
	...
```

Note: my R-b applies with this comment addressed or not)